source/extensions/filters/common/ratelimit/ratelimit_impl.cc

```diff
@@ -0,0 +1,146 @@
+#include "extensions/filters/common/ratelimit/ratelimit_impl.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/api/v2/ratelimit/ratelimit.pb.h"
+#include "envoy/registry/registry.h"
+#include "envoy/singleton/manager.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/assert.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RateLimit {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(ratelimit_client);
+
+ClientFactoryPtr
+ClientFactory ::rateLimitClientFactory(Server::Configuration::FactoryContext& context) {
+  // TODO(ramaraochavali): figure out a way to get the singleton name here.
+  Envoy::RateLimit::RateLimitServiceConfigPtr ratelimit_config =
+      context.singletonManager().getTyped<Envoy::RateLimit::RateLimitServiceConfig>(
+          "ratelimit_config_singleton", [] { return nullptr; });
```

Comment:
@mattklein123  This is the second problem with singleton I am taking about. I have registered the `ratelimit_config` in ConfigurationImpl.cc where we setup the config and added it to singleton there. Now in the extensions, I want to get the already registered singleton object. I could not find a way to get it - I have to hard the singleton name here with a dummy lamda that should never execute if rate limit service config is there. Am I using it correctly? If there is a way to get the rate limit service config set in singleton without hard coding the name - I can remove the method that I added in FactoryContext to get the rateLimitServiceConfig that will remove from Instance.h and Server.h .
If it is ok, to move forward like this also I can remove the rateLimitServiceConfig method from FactoryContext and other places. LMK.

Reply:
Yeah. That is a good option. Instead of creating client from core, I think we have to just register the rate limit config by calling in to extensions. Once the rate limit config is registered, the filter config code can build client using the registered rate limit config because any way the client has to be built there. Also when we introduce rate limit service config in filter, there will be minimal changes because we can also look at the config in the filter and build it, validate it etc. I have made changes like that. PTAL.